internal/objects/service/service.go

```diff
@@ -274,8 +274,23 @@ func DesiredEnvoyService(contour *operatorv1alpha1.Contour) *corev1.Service {
 		}
 	case operatorv1alpha1.NodePortServicePublishingType:
 		svc.Spec.Type = corev1.ServiceTypeNodePort
-		svc.Spec.Ports[0].NodePort = EnvoyNodePortHTTPPort
-		svc.Spec.Ports[1].NodePort = EnvoyNodePortHTTPSPort
+		if contour.Spec.NetworkPublishing.Envoy.NodePorts == nil {
+			// Port numbers will be auto-assigned by API server.
+			httpPort := operatorv1alpha1.NodePort{Name: "http"}
+			httpsPort := operatorv1alpha1.NodePort{Name: "https"}
+			contour.Spec.NetworkPublishing.Envoy.NodePorts = []operatorv1alpha1.NodePort{httpPort, httpsPort}
+		} else {
+			// Note: API server defaulting requires min/max of 2 nodeports.
+			for _, p := range contour.Spec.NetworkPublishing.Envoy.NodePorts {
+				if p.Name == "http" && p.PortNumber != nil {
```

Comment:
this assumes only ports with the correct names come through which could still not happen, re other comments

Reply:
also doesn't this assume the `http` port is first?

by this logic is that always going to be true?

https://github.com/projectcontour/contour-operator/blob/3d5f2ab0c0b8e2fa5a3014d8a418315df54978fa/internal/objects/service/service.go#L216-L238